Avoid "whitelist"

tamird opened the pull request:
Other terms are more inclusive and precise.

tamird committed:
Avoid "whitelist"

Other terms are more inclusive and precise.

rust-highfive commented:
r? @petrochenkov

(rust_highfive has picked a reviewer for you, use r? to override)

lzutao commented on src/ci/docker/host-x86_64/armhf-gnu/vexpress_config:
```diff
@@ -1932,7 +1932,7 @@ CONFIG_USB_ANNOUNCE_NEW_DEVICES=y
 # CONFIG_USB_DYNAMIC_MINORS is not set
 # CONFIG_USB_OTG is not set
-# CONFIG_USB_OTG_WHITELIST is not set
+# CONFIG_USB_OTG_ALLOWLIST is not set
```
?

tamird replied:
Yeah, this one doesn't make sense. Reverted.

ratijas replied:
does _any_ of it make _any_ sense? 🤯

ratijas commented:
WTF

ratijas commented:
Let me expand a little bit on my opinion. I see people upvote/downvote a lot, and I don't want it to happen for a wrong reason.  Lemme get this straight: I'm not a troll; I just genuinely can't believe people are seriously into this fancy tolerastency cowshit.

You even imagine how much it costs to produce reliable software product? Or is it just political toys for you to promote your post on twitter? I, for example, filed half a dozen of issues around NetworkManager/OpenVPN just today — because in IT many things either never worked as expected or just break out of nowhere — and now it will take few weeks to fix 'em and few more years to land onto production systems. Such changes do no good for technology, meanwhile on a social plane, people who _want to_ get offended — they _will_ find their way to get offended by literally anything anyway. I, for instance, am horribly offended by this intolerance against white/black/any other color — why not enjoying the whole RGB palette w/o race, politics or sexual issues? and also by the presence of python scripts in rust repository; so now what — ban python and `sed s/python/ruby/g`? (just kidding, not an adept of ruby either).

I agree with a professor who claimed that in IT things shall not be taken lightly, and such irresponsible initiatives must be met with full severity, classified as intentional injection of harmful code into software, and the whole case taken to the court. Because there have to be an end to it.

You've just changed third-party config variable among other changes. And you proceeded to get pointed out on it. What else did you break? What else did you forget to rename? Who will be responsible for it, huh? You got guts to do programming — why don't you use your energy for the great good™ instead?

tamird commented:
The "third party config variable" was a comment.

cc @steveklabnik and @nikomatsakis, I guess. The reaction to this PR demonstrates the reason this sort of change should be made sooner rather than later.

steveklabnik commented:
r? @rust-lang/compiler

BurntSushi commented:
**Moderation note:** Some of the comments already posted on this PR are completely inappropriate. While dissent is always encouraged, it _must_ be provided in a constructive manner. Mixing it with personal attacks and social commentary is not constructive and it will not be tolerated in official Rust community spaces.

Given the nature of this PR, I am locking this PR with the intent that relevant stakeholders are given time to decide how they want to handle this first. It looks like that's probably @rust-lang/compiler.

petrochenkov commented on src/librustc_feature/builtin_attrs.rs:
```diff
@@ -47,7 +47,7 @@ pub enum AttributeType {
     /// before the unused_attribute check. These attributes
     /// will be ignored by the unused_attribute lint
-    Whitelisted,
+    Allowlisted,
```
This case is responsible for a significant part of renaming, but both the old and new name miss the point.

This should rather be `AssumeUsed` or something (except that the whole enum is pretty strange and should probably be refatored into something else - the crate level restriction and "used" assumptions are orthogonal).

Manishearth replied:
Yeah the usage of Whitelisted here has always bugged me.

A good side benefit of critically evaluating language used for inclusion is that you often end up with more precise language _anyway_ 😁

nikomatsakis replied:
I too was often confused by this existing name and couldn't figure out what it was supposed to mean.

kennytm commented on src/librustc_codegen_ssa/back/linker.rs:
```diff
@@ -1177,7 +1177,7 @@ impl<'a> Linker for WasmLd<'a> {
 
         // LLD will hide these otherwise-internal symbols since our `--export`
-        // list above is a whitelist of what to export. Various bits and pieces
+        // list above is a allowlist of what to export. Various bits and pieces
```
```suggestion
        // list above is an allowlist of what to export. Various bits and pieces
```

kennytm commented on src/librustc_codegen_llvm/attributes.rs:
```diff
@@ -263,7 +263,7 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
     // panic=abort` flag is passed.
     //
-    // You can also find more info on why Windows is whitelisted here in:
+    // You can also find more info on why Windows is allowlisted here in:
```
could we not use "allowlist" as a verb? "allow" or "permit" (as in `src/libcore/ffi.rs`) sounds much better

```suggestion
    // You can also find more info on why Windows is allowed here in:
```

nikomatsakis replied:
This is another case where I think we could be more precise. Can we say:

> You can also find more info on why Windows always requires uwtables here:

I don't really see this as a very clear instance of a allow/deny list or what have you.

kennytm commented on src/test/run-make-fulldeps/sysroot-crates-are-unstable/test.py:
```diff
@@ -5,7 +5,7 @@
 
 
-# This is a whitelist of files which are stable crates or simply are not crates,
+# This is a allowlist of files which are stable crates or simply are not crates,
```
```suggestion
# This is a list of files which are stable crates or simply are not crates,
```

kennytm commented on src/tools/clippy/clippy_lints/src/eq_op.rs:
```diff
@@ -16,7 +16,7 @@ declare_clippy_lint! {
     /// calls (notably [racer](https://github.com/phildawes/racer) had one instance
     /// of `x.pop() && x.pop()`), so we removed matching any function or method
-    /// calls. We may introduce a whitelist of known pure functions in the future.
+    /// calls. We may introduce a allowlist of known pure functions in the future.
```
```suggestion
    /// calls. We may introduce an allowlist of known pure functions in the future.
```

nikomatsakis commented on src/librustc_middle/lint.rs:
```diff
@@ -230,7 +230,7 @@ pub fn struct_lint_level<'s, 'd>(
 
             // If this is a future incompatible lint it'll become a hard error, so
-            // we have to emit *something*. Also allow lints to whitelist themselves
+            // we have to emit *something*. Also allow lints to allowlist themselves
```
I find the use of "allowlist" here also a bit confusing, I think this would be more precisely stated like this?

> Also, if this lint occurs in the expansion of a macro from an external crate, allow individual lints to opt-out from being reported.

nikomatsakis reviewed (approved):
Left a few nits, but I am in favor of removing the term "whitelist" and moving to "allowlist" or other more precise terminology (big 👍 to "AssumeUsed" in particular). This doesn't look like it affects any "user visible" flags or anything like that, it's purely internal? (I would want to have some kind of "deprecation" for the old flags if there were any.)

wesleywiser commented on src/librustc_codegen_llvm/llvm_util.rs:
```diff
@@ -139,7 +139,7 @@ pub fn time_trace_profiler_finish(file_name: &str) {
 // array, leading to crashes.
 
-const ARM_WHITELIST: &[(&str, Option<Symbol>)] = &[
+const ARM_ALLOWLIST: &[(&str, Option<Symbol>)] = &[
```
Nit: It looks like all of these are features, perhaps a better name would be `ARM_FEATURES` or `ARM_TARGET_FEATURES`? 

That would be in keeping with the function below `all_known_features()`.

nikomatsakis replied:
This was bugging me as well -- I didn't like the use of "allowlist" here (I also don't like the use of "whitelist"). I'd prefer `TARGET_FEATURES` and I was thinking that the list of features would be like `target_features_supported` or something like that. That is, this is what we're really looking for, right? What set of features are supported by the current target.

nikomatsakis replied:
(In particular, reading the diff, I found myself confused about the meaning of these lists -- like, what are they listing, and why would something be added to it? I had to read the code a bit to understand what was going on, but if it had had a name like `ARM_TARGET_FEATURES` or even `ARM_SUPPORTED_TARGET_FEATURES` I would have known right away.

wesleywiser commented on src/librustc_codegen_llvm/llvm_util.rs:
```diff
@@ -307,16 +307,16 @@ pub fn target_features(sess: &Session) -> Vec<Symbol> {
 }
 
-pub fn target_feature_whitelist(sess: &Session) -> &'static [(&'static str, Option<Symbol>)] {
+pub fn target_feature_allowlist(sess: &Session) -> &'static [(&'static str, Option<Symbol>)] {
```
Nit: `allowed_target_features` reads a bit better to me.

nikomatsakis replied:
Going by my previous comment, `supported_target_features` might work well.

oli-obk commented on src/bootstrap/doc.rs:
```diff
@@ -439,7 +439,7 @@ impl Step for Std {
             compile::std_cargo(builder, target, compiler.stage, &mut cargo);
 
-            // Keep a whitelist so we do not build internal stdlib crates, these will be
+            // Keep an allowlist so we do not build internal stdlib crates, these will be
             // build by the rustc step later if enabled.
```
```suggestion
            // Avoid building internal stdlib crates, these will be
            // built by the rustc step later if enabled.
```

oli-obk replied:
This comment disappeared entirely, can you reinstate it?

tamird replied:
Done, thanks for the suggestion.

oli-obk commented on src/librustc_codegen_llvm/back/lto.rs:
```diff
@@ -306,7 +306,7 @@ fn fat_lto(
         save_temp_bitcode(&cgcx, &module, "lto.input");
 
-        // Internalize everything that *isn't* in our whitelist to help strip out
+        // Internalize everything that *isn't* in our allowlist to help strip out
```
```suggestion
        // Internalize all symbols that *aren't* exported to help strip out
```

oli-obk commented on src/librustc_codegen_llvm/back/lto.rs:
```diff
@@ -306,7 +306,7 @@ fn fat_lto(
+        // Internalize everything that *isn't* in our allowlist to help strip out
         // more modules and such
         unsafe {
             let ptr = symbol_white_list.as_ptr();
```
Maybe name this `exported_symbol_names` (this variable occurs in several different functions in this file)

oli-obk commented on src/librustc_codegen_ssa/back/linker.rs:
```diff
@@ -1177,7 +1177,7 @@ impl<'a> Linker for WasmLd<'a> {
 
         // LLD will hide these otherwise-internal symbols since our `--export`
-        // list above is a whitelist of what to export. Various bits and pieces
+        // list above is an allowlist of what to export. Various bits and pieces
```
```
        // LLD will hide these otherwise-internal symbols since it only exports
        // symbols explicity passed via the `--export` flags above and hides all
        // others. Various bits and pieces
```

oli-obk commented on src/librustc_expand/base.rs:
```diff
@@ -735,7 +735,7 @@ pub struct SyntaxExtension {
     /// Span of the macro definition.
     pub span: Span,
-    /// Whitelist of unstable features that are treated as stable inside this macro.
+    /// Allowlist of unstable features that are treated as stable inside this macro.
```
```suggestion
    /// List of unstable features that are treated as stable inside this macro.
```

oli-obk commented on src/librustc_metadata/rmeta/decoder.rs:
```diff
@@ -1387,7 +1387,7 @@ impl<'a, 'tcx> CrateMetadataRef<'a> {
             EntryKind::AssocFn(data) => data.decode(self).fn_data.constness,
             EntryKind::Fn(data) => data.decode(self).constness,
             // Some intrinsics can be const fn. While we could recompute this (at least until we
-            // stop having hardcoded whitelists and move to stability attributes), it seems cleaner
+            // stop having hardcoded allowlists and move to stability attributes), it seems cleaner
             // to treat all const fns equally.
```
Uhm... remove this entire comment, it is severely outdated.

oli-obk commented on src/librustc_middle/query/mod.rs:
```diff
@@ -1413,10 +1413,10 @@ rustc_queries! {
             storage(ArenaCacheSelector<'tcx>)
             eval_always
-            desc { "looking up the whitelist of target features" }
+            desc { "looking up the allowlist of target features" }
```
```suggestion
            desc { "looking up the list of allowed target features" }
```

oli-obk commented on src/librustc_mir/const_eval/fn_queries.rs:
```diff
@@ -88,7 +88,7 @@ pub fn is_parent_const_impl_raw(tcx: TyCtxt<'_>, hir_id: hir::HirId) -> bool {
 
 /// Checks whether the function has a `const` modifier or, in case it is an intrinsic, whether
-/// said intrinsic is on the whitelist for being const callable.
+/// said intrinsic is on the allowlist for being const callable.
```
```suggestion
/// said intrinsic has a `rustc_const_{un,}stable` attribute.
```

oli-obk commented on src/librustc_resolve/late/lifetimes.rs:
```diff
@@ -2122,7 +2122,7 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
                     {
                         match path.res {
-                            // Whitelist the types that unambiguously always
+                            // Allowlist the types that unambiguously always
```
```suggestion
                            // Allow the types that unambiguously always
```

oli-obk commented on src/librustc_trait_selection/traits/error_reporting/suggestions.rs:
```diff
@@ -2139,7 +2139,7 @@ pub trait NextTypeParamName {
 impl NextTypeParamName for &[hir::GenericParam<'_>] {
     fn next_type_param_name(&self, name: Option<&str>) -> String {
-        // This is the whitelist of possible parameter names that we might suggest.
+        // This is the allowlist of possible parameter names that we might suggest.
```
```suggestion
        // This is the list of possible parameter names that we might suggest.
```

BurntSushi commented:
**Moderation note:** Let's try unlocking this thread and see how it goes. Folks, please avoid making unconstructive comments. We will be aggressively moderating this PR. If it becomes unwieldy, we will lock the PR again.

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 9a5714db6b6a162e174803a31db97ab53a396d42 has been approved by `oli-obk`

<!-- @bors r=oli-obk 9a5714db6b6a162e174803a31db97ab53a396d42 -->
<!-- homu: {"type":"Approved","sha":"9a5714db6b6a162e174803a31db97ab53a396d42","approver":"oli-obk"} -->

pnkfelix commented:
(I'm nominating because I want this PR to be discussed in tomorrows meeting; however, my nomination should not be interpreted as blocking the PR. I'm fine with it being r+'ed, assuming team feedback has been incorporated.)

ritobanrc commented:
This is actively harmful to improving the material conditions of Black and other disadvantaged and marginalized groups. It's unquestioned that computer science is a field overwhelmingly dominated by white and Asian (incl. Indian, Middle Eastern, etc.) men. Both Black and Latino people, and women, are extremely disproportionately underrepresented in computer science ([citation](https://cra.org/crn/2015/09/expanding-the-pipeline-the-state-of-african-americans-in-computer-science-the-need-to-increase-representation/), [citation](https://ncsesdata.nsf.gov/sere/2018/)), and this is something that we should all strive to improve.  

The fact is, however, that changing whitelist to allowlist does not do that. I can promise you, with 100% certainty, that there has _never_ in the history of the world, ever, been a black person who genuinely, unironically said "I would become a computer science major, if only they'd stop using the term whitelist". There are about a million complicated sociological and economic factors that explain the disparity, and we ought to strive to do our best to combat them. Here are a couple of _genuine_ ways that you could combat them with far more substantive and material impacts:

- Donate (or feature) any one of the dozens of charities that work to improve STEM education in low-income and marginalized communities, like (VoSNL)[https://www.vosnl.org/].
- Join a protest
- Participate in a mutual aid group
- Learn if there is a rent strike or a tenants union in your city
- Read a book by a black author

And dozens more. 

Even just including a statement at the top of the README or on the website saying "Rust stands with people of color and all victims of oppression." with a link to the BLM webpage would do more material good than this PR. 

And yet, despite not making _any_ material improvements to anyone's lives, this PR plays directly into the hands of the alt-right. This only further cements the belief that the "SJWs are speech policing", which is precisely the first step in the alt-right pipeline! Additionally, this furthers the frankly harmful misunderstanding that racial equality is merely removing offensive phrases or whatever, ignoring the far more insidious social factors that play into it.

tamird commented:
@ritobanrc I agree that "allowlist" is not a good term, and it has been expunged from this PR.

protheory8 commented:
Hi, I agree that there were some good changes, like 'target_features_whitelist' to 'supported_target_features'. Allowlist just sounds horrible while whitelist is proper, popular term. I don't see a problem with changing whitelist to other, more precise terms, but I really disagree that whitelist/blacklist should be avoided and I feel like this PR is focused on doing exactly that. Sometimes whitelist is the best term to use and shouldn't be avoided at all.

newpavlov commented:
I also agree that some changes are good, so I would prefer for a different PR to be opened with well motivated changes only. But I strongly disagree with renamings the sole motivations of which is removal of an "offensive" word. I've wrote about `master` branch renaming [here](https://github.com/RustCrypto/meta/issues/2) and I think the post applies in this case as well.

protheory8 commented:
In my opinion whitelist and blacklist are not exclusive terms, they are totally fine, I don't know why would anyone think otherwise honestly.

eminence commented:
Even if you disagree with the motivation, the end result seems to be code that is higher quality and easier to understand (which is a very important thing to consider when reviewing a code change)

protheory8 commented:
Maybe, but I believe there are a few changes where `whitelist` was easier to understand.

ratijas commented:
_Without further heating up the political side of this change_, I shall humbly note that from the etymology's point of view **whitelist/blacklist** are well established terms defined by any English dictionary, while so called "**blocklist**", "**denylist**", "**allowlist**" etc. are compound words invented as a part of recent movements — they do not appear in traditional dictionaries nor in published English textbooks. Even my Firefox spellchecker which claims to be "English (United States, _large_)" underlines them with red wave.

Why am I bringing attention to this? First of all, because there are other countries beyond USA, where people learn English in classes (as opposed to being a mother-tongue). Please, consider twice how much "more precise" are these "other terms" for them.

tamird commented:
@ratijas none of "blocklist", "denylist", "allowlist" are present in this change nor in the other change (#74150).

protheory8 commented:
Really though, is that variable going to be named `EXEMPTIONS`? Really bad name.

ratijas commented:
> @ratijas none of "blocklist", "denylist", "allowlist" are present in this change nor in the other change (#74150).

Good. For the time being, let this comment stay as a reference for the next generations. Feels like, gonna need to copy-paste it a lot in the nearest future.

lnicola commented:
> none of "blocklist", "denylist", "allowlist" are present in this change nor in the other change

Yes, they are:

![image](https://user-images.githubusercontent.com/308347/86955388-24676b00-c160-11ea-9236-5c800e0c9783.png)

protheory8 commented:
Well, OK, why do you think that `EXEMPTIONS` is a better term? It's clearly more difficult to understand. If you want to avoid `WHITELIST`, you should've at least called it `EXCEPTIONS` or something, if I understand correctly.

pnkfelix commented on src/tools/clippy/clippy_lints/src/non_expressive_names.rs:
```diff
@@ -117,7 +117,7 @@ impl<'a, 'tcx> SimilarNamesLocalVisitor<'a, 'tcx> {
 // the assumption is that no name is ever contained in multiple lists.
 #[rustfmt::skip]
-const WHITELIST: &[&[&str]] = &[
+const EXEMPTIONS: &[&[&str]] = &[
```
if you renamed this (and the related occurrence of `exemptions` above) to `ALLOWED_TO_BE_SIMILAR` (and `allowed_to_be_similar`) then the code would be more self-documenting.

pnkfelix replied:
(such a rename may require some extra line-breaking to placate tidy, but I personally would be okay with that.)

tamird replied:
Good point; done.

BurntSushi commented:
@protheory8 You've made your stance clear about `EXEMPTIONS`. Please stop repeating it, especially after your comments have been moderated.

protheory8 commented:
I suggested an alternative in the last one though

tamird commented:
Thanks, updated per suggestions.

TyPR124 commented on src/librustc_passes/weak_lang_items.rs:
```diff
@@ -59,7 +59,7 @@ fn verify<'tcx>(tcx: TyCtxt<'tcx>, items: &lang_items::LanguageItems) {
 
     for (name, &item) in WEAK_ITEMS_REFS.iter() {
-        if missing.contains(&item) && !whitelisted(tcx, item) && items.require(item).is_err() {
+        if missing.contains(&item) && !not_required(tcx, item) && items.require(item).is_err() {
```
"if missing contains (item), and not whitelisted (item)" vs "if missing contains (item), and not not required (item)"

I feel the latter is MUCH harder to interpret than the former. Is "not not required" equivalent to "required" here? If so, could the return value be inverted and name changed usage is more clear?

(even though I see absolutely no problem with the original)

tamird replied:
Done.

BillBarnhill commented:
I agree with the sentiment, but worry that lots of people are doing this and each group is using different names. I wish we would come to a consensus.

For me, I'd prefer stoplight colors. Redlist for stopped, Greenlist for those who can go through, and maybe even Yellow list who can go through but maybe have limited permissions or increased security.

nikomatsakis commented on src/tools/clippy/clippy_lints/src/eq_op.rs:
```diff
@@ -16,7 +16,7 @@ declare_clippy_lint! {
     /// calls (notably [racer](https://github.com/phildawes/racer) had one instance
     /// of `x.pop() && x.pop()`), so we removed matching any function or method
-    /// calls. We may introduce a whitelist of known pure functions in the future.
+    /// calls. We may introduce an list of known pure functions in the future.
```
```suggestion
    /// calls. We may introduce a list of known pure functions in the future.
```

tamird replied:
Done.

nikomatsakis reviewed (changes requested):
Minor nit.

taiki-e commented on src/test/run-make-fulldeps/sysroot-crates-are-unstable/test.py:
```diff
@@ -5,7 +5,7 @@
 
 
-# This is a whitelist of files which are stable crates or simply are not crates,
+# This is an list of files which are stable crates or simply are not crates,
```
```suggestion
# This is a list of files which are stable crates or simply are not crates,
```

tamird replied:
Done.

lnicola commented:
Do these last changes even matter since the PR is already rolled up?

Pzixel commented:
First of all, I'd like to thank you for opening the discussion. It really worth talking about it before making any changes, because it's a precedent after all.

My position still the same, I personally find terms whitelist/blacklist widely accepted across the globe, and renaming things in such way makes it a bit harder to understand meaning of terms. Frensh/Russian/Ukrainian/... and I believe any other European language has this concepts, although they haven't for allowlist and synonyms. I believe it's actually the most precise term for describing anything that should be allowed and blacklist for something that should not. 

I do believe it's more important to make things clearer rather than allow everyday life to affect technology.

So it worth keeping in mind that while it arguably makes things better and more welcoming for US citizens that feel oppressed it makes it worse for multiple non-US citizens that.

---

Being said, it's all very minor and probably doesn't worth all this attention it get but I find this very important junction. I mean I'd like language to not make suboptimal decisions because of anything unrelated. People's views evolve, but languages are not that agile, it's pretty one-way road and once something happens it's likely to remain unchangeable.

lnicola commented:
A policy question for the future: how does the decision to merge this affect future contributions to the language? Will there be lists (ahem) of forbidden words and expressions that people are not allowed to use?

sschueller commented:
If the renaming is purely political it is a mistake and sets a dangerous president. Did you know "Werk-en-Rust" was a slave plantation? What will you do in future if the word Rust becomes "offensive" like a group of people managed to do to the OK sign? What if it's offensive in another language, will those be considered as well?

tamird commented:
I'm happy to address your concerns case-by-case. Without commenting on whether or not the statement that "whitelist/blacklist widely accepted across the globe" is true, the position I'm taking in this PR is that those terms are no longer accepted by me, and the position taken by the compiler team seems to be that the terms I've replaced them with in this PR are better.

Again, I've incorporated all the feedback regarding specific terms that were suboptimal, and I welcome more feedback of that sort. I've done my best to avoid objectively bad substitutes like "allowlist" in this PR.

sschueller commented:
@tamird Can you elaborate why these terms are no longer accepted by you? I would like to understand your view point.

tamird commented:
I'd be happy to, but I won't do it in this venue. Feel free to reach out, I'm not hard to find.

nikomatsakis commented:
I object to the framing of this term as "political", at least in so far as the word "political" is being meant to seem like some kind of cynical action or something related to an election or power. I don't see this PR as being about power or elections. I see this as coming back to the core value that Rust aims to be inclusive and empowering. 

To me, it's as simple as this. There are plenty of people in traditionally under-represented groups who have stated, in good faith, that the existing terms are causing them discomfort and making them feel excluded. If people are genuinely saying that, I think it behooves us to listen and see if we can take some reasonable action to fix that situation. This may not always be possible: e.g., I wouldn't want to break backwards compatibility (though creating synonyms seems fine). But changing the name of some internal variable names seems well within our power, and that is the only question we are being asked to decide now.

As far as general precedent, I don't think this PR creates a firm rule by itself. I do think that the compiler team ought to decide whether to accept new instances of the term whitelist/blacklist. I would be against doing so, and I think we should amend our style guide to make that clear. It's probably a good idea to also clarify what we prefer to do instead, and I think this PR sets good precedent: try to find clearer ways to express the property in question.

As far as historical precedent, obviously the terms whitelist/blacklist have been used for some time. That doesn't mean they don't create problems. People have been passing raw pointers around in C for a long time and yet still we are trying to find something better, after all. =)

nikomatsakis reviewed (approved):
Thanks for fixing those nits, this is looking good.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 7c6a78f4667d3d286537c1e43151394fcb0a28ac has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 7c6a78f4667d3d286537c1e43151394fcb0a28ac -->
<!-- homu: {"type":"Approved","sha":"7c6a78f4667d3d286537c1e43151394fcb0a28ac","approver":"nikomatsakis"} -->

tamird commented:
Minor amendment to fix a whitespace issue.

nikomatsakis commented:
@bors r+

bors commented:
:pushpin: Commit 48d7318d75a6b396e5592579c8cf7042647f82bf has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 48d7318d75a6b396e5592579c8cf7042647f82bf -->
<!-- homu: {"type":"Approved","sha":"48d7318d75a6b396e5592579c8cf7042647f82bf","approver":"nikomatsakis"} -->

yzb3 commented:
I don't doubt that this sort of initiative stems from good intentions, but this is not a viable approach to setting rules.
How would we amend the style guide? By stating that we should avoid words that _might_ be considered offensive in general? Or in the future? This is an impossible rule to follow. The words "whitelist" and "blacklist", just like a few others, have only recently been deemed problematic in some circles; we've used them for a long time and there were no bad intentions behind it.
Or would we create polls asking people about all the words that they find offensive? Who's opinions would matter most? The ones considering themselves underrepresented? An argument is not made more valid depending on who expresses it.

As already mentioned, some of these changes are an improvement, so it's to the benefit of the community. But let's try not to make the original reasoning behind them become a part of governance.

TyPR124 commented:
> There are plenty of people in traditionally under-represented groups who have stated, in good faith, that the existing terms are causing them discomfort and making them feel excluded.

I'm just going to be blunt, because there isn't a nice way to put it:

These people need to accept that "blacklist" and "whitelist" are not racially charged words. They are commonly used, not exclusive to programming or even IT in general. Just looking in the tech world, I could find you a million instances where these words are used and are not going to be changed. If they can't accept these words, they should just bury their head in the sand and ignore the rest of the world, because they are not going to go away.

PS: I'm uncomfortable with double-negatives, especially in code where things are supposed to follow logic. This PR introduces a double negative (which I pointed out earlier), and I don't see that being addressed. Why is the word "whitelist" worse for a codebase than a double-negative in logic flow?

ratijas commented:
> I'd be happy to, but I won't do it in this venue. Feel free to reach out, I'm not hard to find.

CC me pls if you ever gonna write it. My email is also widely posted in public.

> A policy question for the future: how does the decision to merge this affect future contributions to the language? Will there be lists (ahem) of forbidden words and expressions that people are not allowed to use?

It sounds like something that should've been done beforehand. Maybe even RFC with a road-map for adjusting the documentation. As  @nikomatsakis pointed out, this is not a one-time change, but a whole precedence to potentially force new rules on future contributions.

On a **Moderation note**: dead mods, please, stop collapsing peoples' comments. Curious readers would expand them anyway, only with an extra inconvenient click. And per [Streisand effect](https://en.wikipedia.org/wiki/Streisand_effect) such heated discussion tend to make people curious about what was (or rather attempted to be) hidden. Besides, one particular commend quoted above is quite on-topic, IMHO. I would even propose creating a separate thread elsewhere for further off-topic discussions and pinning a link to it somewhere around here.

Manishearth commented:
> How would we amend the style guide?

Literally saying "avoid whitelist and blacklist". If other cases come up, they can be discussed on a case by case basis.

Manishearth commented:
@ratijas please do not discuss moderation actions on the threads being moderated, such comments may be removed. You are free to email the mods directly with your concerns.

SimonImbrogno commented:
> ...there are plenty of people [...] who have stated, in good faith, that the existing terms are causing them discomfort...

I agree that the community should not be causing distress to people, however I feel rather strongly that these kinds of changes are a misstep. This entire discussion is based on the misapprehension that the terms "whitelist" and "blacklist" refer to racial categories when they do not. (And as far as I know, that was never their original meaning!)

If someone is under a genuine misapprehension that a word is being used in a disparaging or discriminatory sense, then the respectful thing to do is to correct the misapprehension and move forward. To act on a claim of this sort when it is knowingly rooted in a falsehood will only serve to leave the complainant with a mistaken belief, and erode one's ability to effectively act on legitimate concerns in the future.

And to the point being made that there are other "more inclusive and precise" terms that can be used. These terms, despite what some people mistakenly believe, are _not racial terms_ and they are _not ambiguous_ when used in context. This is a not a cogent argument.

I do not support disparaging language causing legitimate distress to members of the community, but that's not what this is. This is akin arguing over homophones :/

newpavlov commented:
@nikomatsakis 
>I object to the framing of this term as "political", at least in so far as the word "political" is being meant to seem like some kind of cynical action or something related to an election or power. I don't see this PR as being about power or elections. I see this as coming back to the core value that Rust aims to be inclusive and empowering.

Personally I see it as prioritization of opinions clearly linked to the current political situation in the US. Support of that newspeak-light in the form of marking well-known and widely used words as "no longer accepted" is a political action, as it gives power to people behind it to define which words we can use and what exactly should be considered "inclusive". Worryingly, it's not the first time Rust leadership got influenced by inner US politics, and for those who outside of the supported group it can be seen as a risk for the project and their investments in it.

Yes, current case can be considered relatively minor, but such pressure has a tendency to escalate when given opportunity (see Overton window). What about "dummy" or "black box" widely used across Rust code base? What about accepting contributions from (let's say) an odious Trump supporter who got recently involved in a social network scandal? What about reserving seats in Rust teams for people of color and non-binaries? What about blocking citizens of a country which US will designate as an enemy (which arguably already happens in a soft form)?

SimonImbrogno commented:
@nikomatsakis

> Ironically the author of this PR is a white (straight?) male. Not to deny your words, but can someone provide a good evidence that whitelist/blacklist indeed causes discomfort for people in those groups?

So, maybe the author experiences genuine personal distress over the use of these terms. That's valid and shouldn't be ignored in a community that cares about it's members. I don't think calling the author's legitimacy into question is relevant here.

I'm willing to accept that @tamird is acting in good faith by opening this PR and may genuinely feel that some amount of distress is being inflicted. However, the fact that this distress is rooted in a _misunderstanding over the etymology_, the community should work to _resolve the confusion_ rather than accept the erroneous premise.

nikomatsakis commented:
> This entire discussion is based on the misapprehension that the terms "whitelist" and "blacklist" refer to racial categories when they do not. 

I don't think anyone here has made that claim. Obviously nobody meant to refer to racial categories in the rustc source, and I'm sure the etymological origins of the word go further back. I don't think the original authors were "wrong" to use whitelist/blacklist. People write code and that code later gets changed to make it better. What I would say is that the words clearly have an association of "white == good" and "black == bad", and regardless of their origins, I can understand how they would serve to make people uncomfortable. (To @newpavlov's point, this is not a new request, I can attest that I've seen plenty of tweets about moving away from these sorts of terms; I'm sure if you do a bit of searching you'll find plenty of evidence.)

Look, everybody knows that one of the hardest things about computer science is finding good names for things. Oftentimes, a  name is bad simply because it brings up other, unintended associations. It happens.

SimonImbrogno commented:
@nikomatsakis

> I don't think anyone here has made that claim.

but you later go on to say 

>"white == good" and "black == bad", and regardless of their origins, I can understand how they would serve to make people uncomfortable.

My genuine question would be, If people aren't choosing to interpret the terms "white"/"black" as racial categories, where is the discomfort coming from?

newpavlov commented:
@SimonImbrogno 
>So, maybe the author experiences genuine personal distress over the use of these terms.

I apologize if my post sounded like _ad hominem_, it was not my intent. My point was that I haven't seen a clear indication that whitelist/blacklist is indeed a sufficiently big problem for under-represented groups mentioned by Niko, especially before 2020. Right now situation in US is energyzed, so people are sensitive to words containing "black", but I believe this situation is temporary.

If there is no significant distress for representatives of those groups in long term, why Tamir's current distress should be more significant than for example mine over such forceful renaming?

Manishearth commented:
> My genuine question would be, If people aren't choosing to interpret the terms "white"/"black" as racial categories, where is the discomfort coming from?

Two things can simultaneously be true:

 - the terms whitelist/blacklist do not originate from intents to refer to racial categories. Nobody is saying that their etymology is racist.
 - the terms whitelist/blacklist reaffirm a "white = good, black = bad" analogy that crops up often in our language, which unfortunately overlaps with society's treatment of some racial categories, and is a microaggression that can be avoided.

nikomatsakis commented:
@SimonImbrogno what @Manishearth said :)

@newpavlov Indeed, I was not referring to having seen tweets about other projects taking this step, but rather some of the comments that came up in those conversations. [Here is one such example,](https://github.com/emberjs/rfcs/pull/639#issuecomment-647765471) which also talks about the question of whether this is a problem specific to the US, and, well, indeed many of the themes that arise on this PR.

OK, at this point I think I've said all that I'm able to say, especially tonight, as it's time to go and have dinner.

SimonImbrogno commented:
@nikomatsakis 

Err to be fair, @Manishearth's answer doesn't really address my question. Those connotations would exist even if people didn't use "white" and "black" in reference to skin colour. 

If someone chooses to take that symbolism and connect it to racial terms, that's a racist thing to do. People should not do that. I don't like people who do that.

The fact that someone might do that does not mean the connection exists a priori. :/

I feel at this point we're drifting off topic somewhat though...

ahmedcharles commented on src/librustc_codegen_llvm/attributes.rs:
```diff
@@ -263,7 +263,7 @@ pub fn from_fn_attrs(cx: &CodegenCx<'ll, 'tcx>, llfn: &'ll Value, instance: ty::
     // panic=abort` flag is passed.
     //
-    // You can also find more info on why Windows is whitelisted here in:
+    // You can also find more info on why Windows always requires uwtables here:
```
```suggestion
    // You can also find more info on why Windows always requires unwind tables here:
```

Owez commented:
See https://github.com/rust-lang/rust/pull/74150#issuecomment-655672257

petrochenkov commented:
I have a recent equivalent of this PR from the opposite side of the political spectrum :joy: 
https://www.bbc.com/news/blogs-news-from-elsewhere-53319505
Looks silly, right? Now look into the mirror.
I both cases an inherently neutral color or a set of colors becomes tied to some political construct in the eyes of a vocal minority.

ssokolow commented:
> the terms whitelist/blacklist reaffirm a "white = good, black = bad" analogy that crops up often in our language, which unfortunately overlaps with society's treatment of some racial categories, and is a microaggression that can be avoided.

I've always felt wary of the "whether intentional or unintentional" part of the definition of microaggression and its potential for abuse. Wikipedia says why better than I can:

> A number of scholars and social commentators have criticised the microaggression concept for its lack of scientific basis, over-reliance on subjective evidence, and promotion of psychological fragility. Critics argue that avoiding behaviours that one interprets as microaggressions restricts one's own freedom and causes emotional self-harm, and that employing authority figures to address microaggressions can lead to an atrophy of those skills needed to mediate one's own disputes.[7] Some argue that, because the term "microaggression" uses language connoting violence to describe verbal conduct, it can be (and is) abused to exaggerate harm, resulting in retribution and the elevation of victimhood.[8]

It's certainly true that there are words which shouldn't be used because of their historical context and etymology, but one must be *very* careful to not allow such compassion and respect to spill over into censoring homophones in unrelated contexts.

comex commented:
I'd like to explicitly call out a phenomenon that always seems to happen in these discussions:

Many people seem to have an immediate, visceral negative reaction to being told they should change their use of language.  I can attest to feeling that way myself, not just once but every time I hear a new word called out as problematic.  This is despite the fact that I've long since decided on principle that most efforts to 'deprecate' words are beneficial.

I'm not really sure why this happens.  I think it has to do with feeling like I'm being accused of doing something wrong, for having used that language in the past.

I think people on both sides of the argument could benefit from keeping this phenomenon in mind.  At the risk of being preachy:

People in favor of the change should consider that arguments like 'there's no downside to changing the terminology' are sort of missing the point.

People against the change should consider that their opposition might not be entirely rational.  In particular, with hundreds of thousands of words and gaining new ones every day, the English language is not seriously maimed by losing one once in a while.

ssokolow commented:
> People against the change should consider that their opposition might not be entirely rational. In particular, with hundreds of thousands of words and gaining new ones every day, the English language is not seriously maimed by losing one once in a while.

Normally, I would agree with you, but "black" and "white" are pretty fundamental and pervasive units of metaphor in the English language, and I think contributing to their stigmatization would be unhelpful at best. (In a perfect world, we'd just change how we refer to people with dark brown skin instead.)

oli-obk commented:
>  "black" and "white" are pretty fundamental and pervasive units of metaphor in the English language, and I think contributing to their stigmatization would be unhelpful at best.

That's exactly the issue as I see it. There are several combined words in English and other languages where the use of "black-something" has a negative meaning and "white-something" has a positive meaning. As stated previously, the ethymology is not in question here, but the use of colors for a bad/good decision. Some examples:

* blacklist, a list of things that must be avoided
* whitelist, a list of things that are considered good and only those may be used
* Schwarzarbeit (german: black work): illegal work for which no taxes are paid
* Schwarzgeld (german: black money): money earned without paying taxes
* Schwarzfahren (german: black driving): taking public transport without paying fare
* black market: trading without government supervision (no taxes, illegal goods, illegal obtaining of legal goods)
* blacksite, an superlegal (fancy word for illegal) camp of intelligence agencies
* Weiße Weste (german: white vest): to have a white vest means, to not have done anything bad
* white lie: a lie with good intentions
* white smoke/black smoke: new pope vs the pope has died

Similarly to comex I have a kneejerk reaction to being informed that a word I'm using is problematic to use. The thing I've come to realize is that I'm priviledged to never have had anything about me being used to make me feel bad or discriminate against me. I can't possibly judge objectively what is problematic to use, so I've opted to just listen to ppl that can do so.

> Ironically the author of this PR is a white (straight?) male.

This statement makes me very uncomfortable on several levels.

> Not to deny your words, but can someone provide a good evidence that whitelist/blacklist indeed causes discomfort for people in those groups? Ideally before the current events.

My problem with that is that it's pushing all the heavy lifting on marginalized groups who often times have to live with real consequences for attempting to change something or even just speaking up on a topic. I consider it highly likely that members of marginalized groups are part of the discussion here. Also looking at the moderated comments, I am not surprised at all if loads of ppl are hesitant to speak up or just don't want to or can invest the emotional energy to do so.

I think we should think more about our terminology and am totally fine with questioning everything. It just takes someone to start the discussion, and as this PR has shown, it's enough to start out with a good motivation, even if the implementation isn't ideal. It's the same thing with purely code changes, often a PR changes direction entirely, but keeps the motivation for the change (otherwise you open a different PR).

Lastly I want to state that this entire PR has cleaned up so many things that were using wrong or confusing terminology that I think there wasn't ever any point in using the term to begin with.

TyPR124 commented:
I would just like to explicitly point out that "whitelist" and "blacklist" are used in technology outside of Rust. Rust cannot shield its users from all of these other places where it is used:

[Cisco](https://www.cisco.com/c/en/us/support/docs/security/email-security-appliance/215110-creating-a-whitelist-policy-on-a-cisco-e.html)
[Meraki](https://documentation.meraki.com/zGeneral_Administration/Cross-Platform_Content/Blocking_and_Whitelisting_Clients)
[Juniper](https://www.juniper.net/documentation/en_US/release-independent/sky-atp/topics/reference/general/sky-atp-whitelist-blacklist-overview.html)
[Ubiquiti](https://help.ui.com/hc/en-us/articles/360006893234-UniFi-USG-UDM-Configuring-Internet-Security-Settings#3.2)
[Sophos](https://community.sophos.com/kb/en-us/120432)
[Fortigate](https://docs.fortinet.com/document/fortigate/6.0.0/handbook/594640/fortigate-whitelist)
[Sonicwall](https://www.sonicwall.com/support/knowledge-base/configuring-smtp-real-time-black-list-rbl-filtering-on-the-sonicwall/170505557998744/)
[Palo Alto](https://live.paloaltonetworks.com/t5/minemeld-articles/creating-whitelists/ta-p/72250)
[Forcepoint](https://support.forcepoint.com/KBArticle?id=Why-does-my-black-list-entry-not-work)
[Barracuda](https://campus.barracuda.com/product/emailsecuritygateway/knowledgebase/50160000000IRKdAAO/What+is+the+Barracuda+Whitelist+on+my+Barracuda+Spam+Firewall/)
[Checkpoint](https://sc1.checkpoint.com/documents/R77/CP_R77_ThreatPrevention_WebAdmin/101703.htm#o114370)
[McAfee](https://download.mcafee.com/products/webhelp/4/1033/GUID-B4D9DBDF-A0E5-4302-9ACD-61AFA799B7A3.html)
[Google](https://support.google.com/a/answer/60751?hl=en)
[Amazon](https://aws.amazon.com/blogs/security/tag/whitelisting/)
[Microsoft](https://azure.microsoft.com/en-us/blog/avoiding-ip-blacklisting-with-azure-web-sites/)
[GoDaddy (admittedly... its GoDaddy)](https://www.godaddy.com/help/setup-a-comment-blacklist-in-wordpress-26370)
[NIST](https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-167.pdf)

I've ignored sources that are non-official or community driven. These are all things that are in official vendor documentation (except the NIST one, that's just a whitepaper [according to wikipedia, whitepaper "is an authoritative report or guide that informs readers concisely about a complex issue and presents the issuing body's philosophy on the matter", just so we're clear that it isn't racist either])

cbourjau commented:
First off: I think this discussion has been very interesting and (presumably) well moderated. Thanks! I would like to add one more point which I have, so far, not seen raised in this discussion.

Issues like the one at hand tend to be draining for everybody who gets involved. Given the open nature of this project, it is impossible to avoid such topics from being raised. From a technical point of view the impact of this PR seems to be limited and hardly relevant. Thus, minimizing the damage to the community  and the energy wasted on this issue might be a worthwhile consideration in answering if "white/blacklist" should be removed from the codebase. The reality is that US ("progressive") cultural norms are quite dominant in the Rust community. Fueled by recent events in that part of the world it seems like the number of people who peruse such rewording changes with zealotry has further increased. If "white/blacklist" is not removed from the codebase I believe its likely that similar discussion will just reoccur. As a worst case scenario, I see a real risk of Rust becoming caught in a "shitstorm" eating up yet more resources.

On the other hand, merging this PR is certainly not free of cost to the community either. As mentioned elsewhere in this thread , merging this PR further emphasizes the US centric nature of this project (if this seems contentious I argue that cultural shifts in the EU, China, Russia, or Africa are much less likely to find a Platform in official newsletters or to lead to actual changes in the code base). I'm not saying that this is the end of the world, but it should not be brushed aside either. However, I don't believe that there will be any large "follow-up" costs to the community on this branch, contrary to the "don't merge scenario".

Bottom line: Even though I personally believe that this entire endeavor is counter productive in almost every aspect I'd be in favor of just getting it over it by merging. At the same time I would hope that the Rust project is not going to make a proud, public fuzz about this change. That might further alienate people who feel like the project is being somewhat bullied into making this change.

nikomatsakis commented:
At this point, the PR is r+'d, and I feel that the conversation has reached a natural stopping point. I would like to request that any further comments be limited to details related to merging the PR. If anyone has any further suggestions of specific names that could be improved, there are always follow-up PRs. =)

(If you're on the compiler team and are concerned about this PR merging, feel free to reach out to me privately to discuss.)

pnkfelix commented:
[discussed](https://zulip-archive.rust-lang.org/238009tcompilermeetings/97902weeklymeeting2020070954818.html#203393930) in today's T-compiler meeting. Removing nomination tag.

matu3ba commented:
> The "third party config variable" was a comment.
> 
> cc @steveklabnik and @nikomatsakis, I guess. The reaction to this PR demonstrates the reason this sort of change should be made sooner rather than later.

**Meta**
Avoiding discussion while at the same time attacking the other side (without clarifying a target) is not constructive from your point and has heated up the unnecessary debate.
Any summary on the content and its inappropriateness + short description on the specifics why the words are better in a technical manner would have stopped it at its beginning and also would make better documentation.
This can also be expected for commonly used words to show good intent. Even if you think they are not appropriate.

**Content**
May you explain the difference between the phrase "permit" and "allow" or are they arbitrary chosen as indications of a filter to be applied? This feels inconsistent to me. The opposite of allow is disallow, so this does not feel very fitting.  
Could you change this to either 1. permit and deny or 2.allow and disallow ? I would favor the latter.

BurntSushi commented:
**Moderation note:** I am strictly enforcing @nikomatsakis's request. Meta discussion is off topic, even if your comment contains additional constructive advice on the PR merging. As Niko said, names can be further clarified in follow-up PRs.

matu3ba commented:
1. `permit` has no clear opposite as word, so I would favor `allow` and `disallow` in filters for consistency and simplicity.  
2. `supported` has sometimes the same semantic as `allowed`, so its unclear why the new word is used in a few places.  
3. Instead of `exemptions` the easier word `exclude` can be used.  
4. "explicitly permitted" -> I am not sure permission can be vague in this context

bors commented:
:umbrella: The latest upstream changes (presumably #74195) made this pull request unmergeable. Please resolve the merge conflicts.

oli-obk commented:
@tamird please rebase and fix the conflict.

tamird commented:
@oli-obk done.

oli-obk commented:
@bors r+ p=1 (bitrottyish)

bors commented:
:pushpin: Commit 62cf767a4a39b47677d18110359d9e7152dc9d1c has been approved by `oli-obk`

<!-- @bors r=oli-obk 62cf767a4a39b47677d18110359d9e7152dc9d1c -->
<!-- homu: {"type":"Approved","sha":"62cf767a4a39b47677d18110359d9e7152dc9d1c","approver":"oli-obk"} -->

TundraFizz commented:
This is much ado about nothing.

TundraFizz commented:
This is much ado about nothing.

bors commented:
:hourglass: Testing commit 62cf767a4a39b47677d18110359d9e7152dc9d1c with merge 2bba01db2eec0bab687b341af01c99cb1e7341fd...
<!-- homu: {"type":"BuildStarted","head_sha":"62cf767a4a39b47677d18110359d9e7152dc9d1c","merge_sha":"2bba01db2eec0bab687b341af01c99cb1e7341fd"} -->

TundraFizz commented:
This is much ado about nothing.

TundraFizz commented:
> Even if you disagree with the motivation, the end result seems to be code that is higher quality and easier to understand (which is a very important thing to consider when reviewing a code change)

The variable name `whitelist` was changed to `listed`. This objectively makes it lower quality and harder to understand.

Manishearth commented:
@bors retry yield

TundraFizz commented:
This is much ado about nothing.

TundraFizz commented:
> Even if you disagree with the motivation, the end result seems to be code that is higher quality and easier to understand (which is a very important thing to consider when reviewing a code change)

The variable name whitelist was changed to `listed`. This objectively makes it lower quality and harder to understand.

lzutao commented:
There is no need to spam like that @TundraFizz

TundraFizz commented:
This is much ado about nothing.

TundraFizz commented:
> Even if you disagree with the motivation, the end result seems to be code that is higher quality and easier to understand (which is a very important thing to consider when reviewing a code change)

The variable name whitelist was changed to `listed`. This objectively makes it lower quality and harder to understand.

lzutao commented on src/etc/test-float-parse/runtests.py:
```diff
@@ -195,9 +195,9 @@ def main():
-    whitelist = sys.argv[1:]
-    if whitelist:
-        tests = [test for test in tests if test in whitelist]
+    listed = sys.argv[1:]
```
`listed` sound like a bool to me. Could we change it to `args` or `arguments` ?

shirshak55 replied:
yea listed is abit weird word here args seems reasonable.

SomeRandomGuy81 commented:
What the hell? Instead of doing something useful to compiler you guys create PR like this? Translating white/black connotation to skin color is racism in its purest form.

BurntSushi commented:
Now that the PR has been merged and this PR seems to have attracted trolls, I'm going to lock it for good. As Niko said above, name improvements can be iterated on in follow up PRs.

Thank you to everyone who participated in this PR constructively!
